sycl/doc/extensions/experimental/sycl_ext_oneapi_cuda_tex_cache_read.asciidoc

```diff
@@ -0,0 +1,110 @@
+= sycl_ext_oneapi_cuda_tex_cache_read
```

Comment:
Adding a global comment here, so the conversation can be threaded.

Is this function supported for all devices, or only for NVIDIA ones?  If it is supported only for NVIDIA devices, is it supported on all of them or only certain models?

One option is to support the function for all devices but implement it as a simple load for those that don't have this special tex cache.

The other option is to treat it as an "optional kernel feature".  This means that there must be some associated aspect which applications can test to see if the device supports the feature.  If it is supported on all Nvidia devices, we could have an aspect like `aspect::ext_oneapi_nvidia` or `aspect::ext_oneapi_cuda`.

Reply:
Also wrt an aspect that could be used for backend specific features: sounds like a good idea to me generally for cuda only/ intel only extensions; For this particular feature I don't think it would work. I think you would need a preprocessor check such as e.g.

```
#ifdef EXT_ONEAPI_CUDA
auto am_i_tex_cache = __ldg(address)
#else
auto am_i_tex_cache = *address.
```

Reply:
In the code snippet you show above with the `#ifdef`, I assume you mean that the application can test the extension's feature-test macro `SYCL_EXT_ONEAPI_CUDA_TEX_CACHE_READ`?

This won't work because the feature-test macro is always defined by DPC++, for all compiler passes.  Therefore, an application that wants to mix CUDA kernels with non-CUDA kernels in the same source file wouldn't be able to use this macro to conditionalize device code as you show.

We have two basic design patterns for dealing with device- or backend-specific APIs in kernel code.  One option is to support the API on all devices / backends.  In your case, this would mean that `__ldg` could be implemented as `return *address` on non-NVIDIA devices.  The second option is to treat the API as an "optional kernel feature" as I describe above and tie it to some aspect (or to one of the `architecture` enums).  In this second case, the application must do one of two things:

* The host code can test the device and only submit the kernel to the device if it has the required aspect, or
* The device code can use the experimental `if_architecture_is` function to call `__ldg` conditionally based on the device architecture.  (This only works for AOT currently.)

Reply:
OK, I think it makes sense to use whatever option is consistent with how users are expected to use similar optional features generally in SYCL. Do you think it is best to implement `__ldg` on all devices but just `return *address` (This is actually what happens in HIP: HIP AMD does a no op for `__ldg`) or the other option?

Reply:
I think it would probably be easier for application code if this API was implemented everywhere, and just did `return *address` as a fallback.

The spec should be reworded to make it clear that this function is available for all devices.  The Overview section should also note that the "ld-global-nc ptx instruction" is specific to CUDA (which might not be obvious to a non-CUDA programmer).

Some other statements I noticed:

> The data is cached in the read-only texture cache

Should probably say something like "If the kernel runs on a CUDA device, the data is cached in the read-only texture cache".

> This function is not available in host code

Should this restriction be removed?

Reply:
> 
> Should this restriction be removed?

I'm not sure, I think it depends on whether there is going to be a host device supported. Do you know whether this is the case?

It certainly should not be available in host code outside of a kernel.

All the other suggestions you made sound good and I will apply them. One thing I am uncertain about is the naming of this extension; should it remain "SYCL_EXT_ONEAPI_CUDA_TEX_CACHE_READ" or switch to "SYCL_EXT_ONEAPI_TEX_CACHE_READ" once I implement `__ldg` on all devices, even if only the CUDA case will be using the tex cache?

Reply:
> It certainly should not be available in host code outside of a kernel.

My question was about allowing `__ldg` outside of a kernel.

> should it remain "SYCL_EXT_ONEAPI_CUDA_TEX_CACHE_READ" or switch to "SYCL_EXT_ONEAPI_TEX_CACHE_READ"

If we are going to use the function name `__ldg`, then I think the current extension name with "CUDA" makes sense.  The name "__ldg" only makes sense to a CUDA programmer.

Longer term, I think we might want to use the `annotated_ptr` class to convey this caching information instead.  The proposal for [sycl_ext_oneapi_annotated_ptr](https://github.com/intel/llvm/pull/5755) has been in review for a long time, though, and I presume you do not want to wait for that to be finished.  I'm thinking that sycl_ext_oneapi_cuda_tex_cache_read could stay as an experimental extension until `annotated_ptr` is ready.

Actually, this would be a good thing to capture as an issue in your extension spec.  Could you add an Issues section at the bottom saying something like:

> Investigate exposing this functionality through `annotated_ptr` once the sycl_ext_oneapi_annotated_ptr becomes finalized.

Reply:
> If we are going to use the function name `__ldg`, then I think the current extension name with "CUDA" makes sense. The name "__ldg" only makes sense to a CUDA programmer.
> 
OK makes sense. Note that I have also kept to the `experimental::cuda::__ldg` namespace.

> Longer term, I think we might want to use the `annotated_ptr` class to convey this caching information instead. The proposal for [sycl_ext_oneapi_annotated_ptr](https://github.com/intel/llvm/pull/5755) has been in review for a long time, though, and I presume you do not want to wait for that to be finished. I'm thinking that sycl_ext_oneapi_cuda_tex_cache_read could stay as an experimental extension until `annotated_ptr` is ready.
> 
> Actually, this would be a good thing to capture as an issue in your extension spec. Could you add an Issues section at the bottom saying something like:
> 
> > Investigate exposing this functionality through `annotated_ptr` once the sycl_ext_oneapi_annotated_ptr becomes finalized.

Sounds good. I've added this note.

I also added a new section "Backend support status". We were thinking of making a PR updating the extensions/template.asciidoc to include this section in order to formalize some of the points you made here: requiring the author to state:

- Which backends this extension is currently supported on
- Which backends could this extension support in the future (at least an initial investigation of this)
- If the extension is not supported by all backends then how do we achieve portable code (is the extension implemented as no-op such as with `__ldg`, do we depend on device aspects, MACROs etc)?

What do you think?

We were also wondering whether it should become standard procedure for new extensions targetting a single backend to be reviewed by someone who is mainly working on a different backend to the one that the extension supports: at least so that those working on other backends can provide additional information on whether the extension might be extendable to other backends. Similar to how I needed to confirm that there wasn't an analogue to `__ldg` in level_zero. What do you think?

Reply:
Adding the "Backend support status" section seems reasonable, and I do think it is good to raise awareness of the need to address code portability when an extension maps primarily to a specific backend.  Depending on how lengthy a description you want to write, you could add some new description to [README-process](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md) in addition to the new "Backend support status" section in the template.

Regarding review by people doing other backends ... I'm certainly in favor of wider reviews.  In the past, we usually expect the person proposing an extension to be responsible for understanding its impact on all backends.  Of course, that person could consult with other people if they are not familiar with all backends.  Do you think there should be some more formal process?